pkg/maglev/maglev.go

```diff
@@ -89,90 +147,123 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
 	// ignore the returned error from wp methods. Also as our task func never
 	// return any error, we have no use returned value from Drain() and don't
 	// need to provide an id to Submit().
-	wp := workerpool.New(numCPU)
-	defer wp.Close()
 	for g := 0; g < bCount; g += batchSize {
 		from, to := g, g+batchSize
 		if to > bCount {
 			to = bCount
 		}
-		wp.Submit("", func(_ context.Context) error {
+		ml.wp.Submit("", func(_ context.Context) error {
 			for i := from; i < to; i++ {
-				offset, skip := getOffsetAndSkip(backends[i], m)
-				permutation[i*int(m)] = offset % m
-				for j := uint64(1); j < m; j++ {
-					permutation[i*int(m)+int(j)] = (permutation[i*int(m)+int(j-1)] + skip) % m
+				h1, h2 := murmur3.Hash128(backends[i].addrBytes, ml.seedMurmur)
+				offset := h1 % uint64(m)
+				skip := (h2 % (uint64(m) - 1)) + 1
+				start := i * m
+				ml.permutations[start] = offset
+				for j := 1; j < m; j++ {
+					ml.permutations[start+j] = (ml.permutations[start+(j-1)] + skip) % uint64(m)
 				}
 			}
 			return nil
 		})
 	}
-	wp.Drain()
+	ml.wp.Drain()
 
-	return permutation[:bCount*int(m)]
+	return ml.permutations[:bCount*int(m)]
 }
 
-// GetLookupTable returns the Maglev lookup table of the size "m" for the given
-// backends. The lookup table contains the IDs of the given backends.
+type backendInfo struct {
+	id        loadbalancer.BackendID
+	addrBytes []byte
+	weight    uint64
+}
+
+type Backend interface {
+	WeightInfo() (id loadbalancer.BackendID, addr loadbalancer.L3n4Addr, weight uint16)
+}
+
+func getBackendInfos[BE Backend](ml *Maglev, backends iter.Seq[BE]) []backendInfo {
+	ml.backendInfos = ml.backendInfos[:0]
+	for b := range backends {
```

Comment:
This is a little confusing. The interface Backend returns the entire object state at least at this juncture.  Thus why not just export and return a BackendInfo?
Also the func name is misleading since we are actually `re-setting` backends and returning them.

Reply:
FYI: This part of the PR will most likely be moved to a separate PR.

Again, this was authored by @joamaki, but let me share my perspective.

> Also the func name is misleading since we are actually re-setting backends and returning them.

`ml.backendInfos` is an un-exported field of type `Maglev` and it is clearly marked as a reusable buffer in the type definition, so I think it's fine to modify it "silently". It would be more natural to do it in a method than in a standalone function, but I guess it is impossible with generics, right?

> The interface Backend returns the entire object state at least at this juncture. Thus why not just export and return a BackendInfo?

I believe the point is to use `GetLookupTable[Be Backend]` with different actual types `BE` implementing the `maglev.Backend` interface. It is used with `experimental.maglevBackend` in `pkg/loadbalancer/experimental/bpf_reconciler.go` and with `*loadbalancer.Backend` in `pkg/maps/lbmap/lbmap.go`.

Reply:
I think I'll refactor this code to have `type BackendInfo struct` and move `GetLookupTable` back into a method and take `iter.Seq[BackendInfo]`. This also allows us to avoid the allocations and temporary data structures, but it's much more direct.

Reply:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.